# ------------------------------------------------------------------------------
#title: "TCR Clonotype Analysis from Single-Cell RNA-seq"
#Author: Nasim Rahmatpour
#Date: "8/12/2022"
# ------------------------------------------------------------------------------



# Project Overview

#This analysis aims to characterize T cell clonal diversity by parsing V(D)J sequencing results and integrating them with scRNA-seq data. The input comes from the `all_contig_annotations.csv` file generated by the Cell Ranger VDJ pipeline. TCR alpha (TRA), beta (TRB), and paired alpha-beta (TRAB) chains are extracted, saved, and annotated for each cell barcode.

#After integration with Seurat RNA data, cells are categorized as:
#- **no_TCR**: cells with no detected TCR  
#- **unique_TCR**: cells with a private (non-shared) clonotype  
#- **clonal_TCR**: cells that share the same clonotype (expanded clones)

#UMAP plots are used to visualize these clonotype categories across the transcriptomic landscape.

---

###Task 1: Parsing and Saving TCR Chain Information
  auto_basic <- function(TCR_List = NA,
                         Sample_List = NA,
                         high_confidence="true",
                         productive="true",
                         output="./NGS067_TCR_outputs/"){
    
    library(dplyr,verbose = F)
    
    TCR_PATH = TCR_List
    Sample_List = Sample_List
    
    high_confidence_p = high_confidence
    productive_p = productive
    output_parameter = output
    
    
    for (i in 1:length(TCR_PATH)) {
      tcr_tmp <- read.csv(paste0(TCR_PATH[i]))
      tcr_tmp$barcode <- paste0("Data_",Sample_List[i],"_",tcr_tmp$barcode)
      assign(x = paste0("tcr_data_",Sample_List[i]),value = tcr_tmp)
    }
    
    if (length(Sample_List) == 1) {
      m <- function(x){eval(as.name(x))}
      tcr_data <- m(ls(pattern="tcr_data_"))
    }else {
      m <- function(x){eval(as.name(x))}
      the_tcr <- ls(pattern="tcr_data_")[1]
      other_tcr <- ls(pattern="tcr_data_")[-1]
      tcr_data <- m(the_tcr)
      for (l in other_tcr) {
        tcr_data <- rbind(tcr_data,m(l))
      }
      
    }
    
    
    tcr_data <- subset(tcr_data,high_confidence == "True"|high_confidence == "true")
    tcr_data <- subset(tcr_data,productive == "True"|productive == "true")
    
    tcr_data$Clonotype <- tcr_data$cdr3_nt
    
    #Subset_TRA TRB TRG TRD
    for(x in unique(tcr_data$chain)){
      assign(x, subset(tcr_data,chain == x) %>% group_by(barcode) %>% arrange(desc(umis)) %>% dplyr::slice(1:1))
    }
    if(exists("TRA")==T){
      write.csv(x = TRA,file = paste0(output_parameter,"TCR_RawData_TRA.csv"))
    }else{
      #print("No TRA")
    }
    
    if(exists("TRB")==T){
      write.csv(x = TRB,file = paste0(output_parameter,"TCR_RawData_TRB.csv"))
    }else{
      #print("No TRB")
    }
    
    if(exists("TRG")==T){
      write.csv(x = TRG,file = paste0(output_parameter,"TCR_RawData_TRG.csv"))
    }else{
      #print("No TRG")
    }
    
    if(exists("TRD")==T){
      write.csv(x = TRD,file = paste0(output_parameter,"TCR_RawData_TRD.csv"))
    }else{
      #print("No TRD")
    }
    
    if (exists("TRA")==T & exists("TRB")==T) {
      TRAB <- inner_join(TRA, TRB, by = "barcode",suffix = c(".TRA", ".TRB"))
      TRAB$Clonotype <-  paste(TRAB$cdr3_nt.TRA, TRAB$cdr3_nt.TRB, sep = "-")
      TRAB$chain = "TRAB"
      write.csv(x = TRAB,file = paste0(output_parameter,"TCR_RawData_TRAB.csv"))
    }else{
      #print("No TRAB")
    }
    
    if (exists("TRG")==T & exists("TRD")==T) {
      TRGD <- inner_join(TRG, TRD, by = "barcode",suffix = c(".TRG", ".TRD"))
      TRGD$Clonotype <-  paste(TRGD$cdr3_nt.TRG, TRGD$cdr3_nt.TRD, sep = "-")
      TRGD$chain = "TRGD"
      write.csv(x = TRGD,file = paste0(output_parameter,"TCR_RawData_TRGD.csv"))
    }else{
      #print("No TRGD")
    }
  }
  
  TCRList <- c("NGS067_TCR_1","NGS067_TCR_2")
  
  InputDirList <- c("./NGS067_TCR_1/all_contig_annotations.csv",
                    "./NGS067_TCR_2/all_contig_annotations.csv")
  
  #calling the fuction, makes 3 csv file
  
  bird_auto_basic(TCR_List = InputDirList,
                  Sample_List = TCRList,
                  output = "./NGS067_TCR_outputs/")
  
###Task 2: Integration with Seurat and Visualization
  data = `4-Data_Annotation_new`
  
  library(dplyr)
  VDJChainList <- list.files("./NGS067_TCR_outputs/",pattern = "TCR_RawData_")
  nColMetadata <- ncol(data@meta.data) + 1
  for (i in VDJChainList) {
    DF_Barcode <- as.data.frame(row.names(data@meta.data))
    colnames(DF_Barcode) <- "barcode"
    
    VDJ_Tmp <- read.csv(paste0("./NGS067_TCR_outputs/",i))
    VDJ_Tmp$barcode <- gsub("TCR_", "", gsub("Data_NGS067_", "Data_NGS067_GEX_", VDJ_Tmp$barcode))
    
    ChainType <- as.character(unique(VDJ_Tmp$chain))
    DF_Barcode <- left_join(DF_Barcode, VDJ_Tmp, by = "barcode")
    DF_Barcode[is.na(DF_Barcode)] <- "0"
    data$tmp <- DF_Barcode$Clonotype
    colnames(data@meta.data)[nColMetadata] <- paste0("ClonoType",ChainType)
    nColMetadata <- nColMetadata+1
  }
  ExistTRAB <- length(grep(pattern = "TRAB",x = colnames(data@meta.data),value = TRUE)) == 1
  data$ClonoTypeTCR <- data$ClonoTypeTRAB
  
  MetaData <- data@meta.data
  MetaData$ClonoTypeTCR <- as.character(MetaData$ClonoTypeTCR)
  MetaData[is.na(MetaData)] <- "0"
  data$ClonoTypeTCR <- MetaData$ClonoTypeTCR
  data$ClonoTypeTRA <- NULL
  data$ClonoTypeTRAB <- NULL
  data$ClonoTypeTRB <- NULL
  
  TCR_Count <- as.data.frame(table(data$ClonoTypeTCR))
  TCR_Count$CellCount <- ifelse(TCR_Count$Var1 == 0,yes = 0,no = TCR_Count$Freq)
  colnames(TCR_Count) <- c("ClonoTypeTCR","CountC","TCRClonoTypeCount")
  MetaData <- data@meta.data
  library(dplyr)
  MetaData_Tmp <- left_join(MetaData,TCR_Count,"ClonoTypeTCR")
  data$TCRClonoTypeCount <- MetaData_Tmp$TCRClonoTypeCount
  
  data$TCRCoverageState <- ifelse(test = data$TCRClonoTypeCount == 0,
                                  yes = "NO-TCR",
                                  no = ifelse(test = data$TCRClonoTypeCount == 1,
                                              yes = "Unique",
                                              no = "Clonal")
  )
  table(data$TCRCoverageState)
  
  P_T <- DimPlot(data,group.by = "TCRCoverageState",pt.size = 1)+coord_fixed()
  pdf(paste0("./NGS067_TCR_outputs//TCRCoverageState.pdf"),width = 12,height = 9)
  print(P_T)
  dev.off()
  
  df_tmp <- as.data.frame.matrix(table(data$T_subtypes_Annotation,data$TCRCoverageState))
  openxlsx::write.xlsx(df_tmp,file = "./NGS067_TCR_outputs//CellCount_TCRCoverageState.xlsx",row.names = TRUE)  